code/fcn_2D_segm/train_fcn8.py

```diff
@@ -0,0 +1,417 @@
+
+    # Apply mask
+    acc *= mask
+    acc = T.sum(acc) / T.sum(mask)
+
+    return acc
+
+
+def crossentropy_metric(y_pred, y_true, void_labels, one_hot=False):
+    # Clip predictions
+    y_pred = T.clip(y_pred, _EPSILON, 1.0 - _EPSILON)
+
+    if one_hot:
+        y_true = T.argmax(y_true, axis=1)
+
+    # Create mask
+    mask = T.ones_like(y_true, dtype=_FLOATX)
+    for el in void_labels:
+        mask = T.set_subtensor(mask[T.eq(y_true, el).nonzero()], 0.)
+
+    # Modify y_true temporarily
+    y_true_tmp = y_true * mask
+    y_true_tmp = y_true_tmp.astype('int32')
+
+    # Compute cross-entropy
+    loss = T.nnet.categorical_crossentropy(y_pred, y_true_tmp)
+
+    # Compute masked mean loss
+    loss *= mask
+    loss = T.sum(loss) / T.sum(mask)
+
+    return loss
+
+
+SAVEPATH = 'save_models/'
+LOADPATH = SAVEPATH
+WEIGHTS_PATH = SAVEPATH
+
+
+def train(dataset, learn_step=0.005,
+          weight_decay=1e-4, num_epochs=500,
+          max_patience=100, data_augmentation={},
+          savepath=None, #loadpath=None,
+          early_stop_class=None,
+          batch_size=None,
+          resume=False,
+          train_from_0_255=False):
+
+    #
+    # Prepare load/save directories
+    #
+    exp_name = 'fcn8_' + 'data_aug' if bool(data_augmentation) else ''
+
+    if savepath is None:
+        raise ValueError('A saving directory must be specified')
+
+    savepath = os.path.join(savepath, dataset, exp_name)
+    # loadpath = os.path.join(loadpath, dataset, exp_name)
+    print savepath
+    # print loadpath
+
+    if not os.path.exists(savepath):
+        os.makedirs(savepath)
+    else:
+        print('\033[93m The following folder already exists {}. '
+              'It will be overwritten in a few seconds...\033[0m'.format(
+                  savepath))
+
+    print('Saving directory : ' + savepath)
+    with open(os.path.join(savepath, "config.txt"), "w") as f:
+        for key, value in locals().items():
+            f.write('{} = {}\n'.format(key, value))
+
+    #
+    # Define symbolic variables
+    #
+    input_var = T.tensor4('input_var')
+    target_var = T.ivector('target_var')
+
+    #
+    # Build dataset iterator
+    #
+    if batch_size is not None:
+        bs = batch_size
+    else:
+        bs = [10, 1, 1]
+    train_iter = Polyps912Dataset(which_set='train',
+                                  batch_size=batch_size[0],
+                                  seq_per_subset=0,
+                                  seq_length=0,
+                                  data_augm_kwargs=train_data_augm_kwargs,
+                                  return_one_hot=one_hot,
+                                  return_01c=False,
+                                  overlap=0,
+                                  use_threads=False,
+                                  shuffle_at_each_epoch=shuffle_train,
+                                  return_list=True,
+                                  return_0_255=return_0_255)
```

Comment:
The train loop has a data_augmentation argument, which is defined below with
- default={'crop_size': (224, 224), 'horizontal_flip': True, 'fill_mode':'constant'},
   
This assure that all images have the same shape (224,224) by using cropping. Also, it enables using minibatch size >1.
If not cropped, then we must use batch_size = [1,1,1], otherwise we can use batch_size=[10,1,1] (for example)